spec/requests/serialization_spec.rb

```diff
@@ -33,43 +33,92 @@
+    expect { Incident.from_hash(JSON.parse(@invalid_json), @user) }.to raise_error do |error|
       expect(error.message).to include('Validation failed')
       expect(error.message).to include('[Screener] Shots fired Can\'t be blank')
       expect(error.message).to include("[General info] Incident date str invalid year")
       expect(error.message).to include('[Involved civilian] Gender is not included in the list')
     end
   end
 
-  it 'cries on broken JSON' do
-    straight_up_broken_json = @valid_json[5..-1]
-    expect { Incident.from_json(straight_up_broken_json, @user) }.to raise_error("JSON parsing error: invalid JSON")
+  it 'cries when given incorrect data' do
+    # Note that, while this test uses JSON clarity for simplicity,
+    # the same results hold for XML import, since the method being tested,
+    # Incident.from_hash, operates on hashes that have already been
+    # parsed from JSON or XML.
 
     json_with_missing_fields = {
       "ori" => @user.ori,
       "screener" => JSON.parse(@valid_json)['screener']
     }.to_json
-    expect { Incident.from_json(json_with_missing_fields, @user) }.to raise_error("JSON parsing error: missing section: general_info")
+    expect { Incident.from_hash(JSON.parse(json_with_missing_fields), @user) }.to raise_error("Parsing error: missing section: general_info")
 
     json_with_extra_garbage = @valid_json.sub('"shots_fired":"t"', '"shots_fired":"t","secondary_agency":"t"')
-    expect { Incident.from_json(json_with_extra_garbage, @user) }.to raise_error("JSON parsing error: unexpected field: secondary_agency in Screener")
+    expect { Incident.from_hash(JSON.parse(json_with_extra_garbage), @user) }.to raise_error("Parsing error: unexpected field: secondary_agency in Screener")
   end
 
-  it 'can create an incident via the JSON upload form', driver: :poltergeist do
-    # TODO: Test that ORI validation is enabled in devise/siteminder mode and disabled in demo mode.
+  describe '[Bulk upload]', driver: :poltergeist do
+    let(:valid_xml) do
+      { incident: JSON.parse(@valid_json) }.to_xml(root: 'incidents')
+                                           .sub('<incidents>', '<incidents type="array">')
+    end
+
+    it 'can create an incident via JSON upload' do
+      json_file = Tempfile.new(['incident', '.json'])
+      json_file.write("[ #{@valid_json} ]")
+      json_file.close
+
+      visit 'incidents/upload'
+      execute_script('$("input[type=file]").show().appendTo("form");')
+      attach_file('file', json_file.path)
+
+      expect(page).to have_content('Uploading 1 incident ...')
+      expect(page).to have_content('Created incident')
+      expect(Incident.count).to eq(1)
+      expect(Incident.first).to be_valid
+      expect(Incident.first).to be_approved
+    end
+
+    it 'displays an error message on broken JSON' do
+      json_file = Tempfile.new(['incident', '.json'])
+      json_file.write("[ #{@valid_json} ]}")
```

Comment:
OK It took me a lonnnnnng time to see how this was different from the previous example. Can you please either add a comment or make it more obvious? Like append `<blah>` as in the other example?